### PR TITLE
search & sort

### PR DESCRIPTION
**Pull request by anastasiiahryniv**
- [x] search
- [x] sort
- [x] course 
- [x] pundit
- [x] n+1
- [x] simplecov min value 
- [x] servises
- [x] decoratos

Screenshot 2024-11-04 at 00 04 03
Screenshot 2024-11-04 at 00 04 54
Screenshot 2024-11-04 at 00 05 33
Screenshot 2024-11-04 at 00 06 03

### CODE REVIEW

**Review by VasylYaremkiv — CHANGES_REQUESTED** [review]

File: app/policies/course_policy.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class CoursePolicy < ApplicationPolicy
+  class Scope < Scope
+    def resolve
+      if user.is_a?(Instructor) && user.instructor_admin?
+        scope.all
+      elsif user.is_a?(Instructor) && user.instructor?
+        scope.where(instructor_id: user.id)
+      elsif user.is_a?(Student)
+        scope.all
```
Comment (VasylYaremkiv) [inline app/policies/course_policy.rb:9]: scope.active

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -1,24 +1,27 @@
 class CoursesController < ApplicationController
   before_action :set_course, only: %i[show edit update destroy]
-  before_action :authenticate_instructor!
```
[the remark this answers is not in the document]
Comment (anastasiiahryniv) [reply app/controllers/courses_controller.rb:3]: https://github.com/anastasiiahryniv/learning_system/pull/9/files#diff-8601de36d6f3a7c68855de1c7ba7bd26d9d71793c8ce315b844959aa6278a7ddR1

File: app/controllers/courses_controller.rb [hunk 3]
```diff
@@ -18,28 +18,15 @@ def edit
   end
 
   def create
-    @course = current_instructor.courses.build(course_params)
```
Comment (VasylYaremkiv) [inline app/controllers/courses_controller.rb:21]: Service should not know about controller.
```
service = CourseCreationService.new(course_params, current_instructor)
if service.call
  @course = service.course
  flash[:notice] = I18n.t('course_create_success')
  redirect_to course_path(@course)
else
  flash[:notice] = I18n.t('course_create_failed')
  render :new
end
```

File: app/controllers/courses_controller.rb [hunk 4]
```diff
@@ -18,28 +18,15 @@ def edit
...
+    CourseCreationService.new(course_params, current_instructor.id, self).create_course
   end
 
   def update
-    if @course.update(course_params)
-      redirect_to course_path(@course)
-    else
-      render 'edit'
-    end
+    CourseUpdatingService.new(@course, course_params, self).update_course
```
Comment (VasylYaremkiv) [inline app/controllers/courses_controller.rb:25]:
```
service = CourseUpdatingService.new(@course, course_params)
if service.call
  flash[:notice] = I18n.t('course_update_success')
  redirect_to course_path(@course)
else
  flash[:notice] = I18n.t('course_update_failed')
  render :edit
end
```

File: app/services/course_creation_service.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  private
+
+  def course
+    Course.new(@params.merge(instructor_id: @instructor_id))
+  end
+
+  def mailer(sender_id, course_id)
```
Comment (VasylYaremkiv) [inline app/services/course_creation_service.rb:27]: rename to `send_mail`

File: app/services/course_creation_service.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+class CourseCreationService
+  def initialize(params, instructor_id, controller)
```
Comment (VasylYaremkiv) [inline app/services/course_creation_service.rb:2]: attr_reader :params, :instructor, :course

File: app/services/course_creation_service.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    @params = params
+    @instructor_id = instructor_id
+    @controller = controller
+  end
+
+  def create_course
+    course.tap do |c|
```
Comment (VasylYaremkiv) [inline app/services/course_creation_service.rb:9]:
```
@course = instrctor.courses.create(params)
send_mail if cource.valid?

cource.valid?
```

**Review by VasylYaremkiv — CHANGES_REQUESTED** [review]

File: app/views/courses/index.html.erb [hunk 8]
```diff
@@ -1,32 +1,39 @@
...
+        <th scope="col"><%= sort_link(@q, :status, 'Status', default_order: :asc) %> </th>
+        <th scope="col"><%= sort_link(@q, :students_count, 'Students count', default_order: :asc) %> </th>
+        <th scope="col">Links</th>
+      </tr>
+    </thead>
+    <tbody>
+      <% @courses.each do |course| %>
+        <tr>
+          <td><%= course.name.truncate(40)%></td>
+          <td><%=course.description.truncate(40) %></td>
```
Comment (VasylYaremkiv) [inline app/views/courses/index.html.erb:19]: Use decorators: https://medium.com/@rwubakwanayoolivier/using-the-decorator-pattern-in-ruby-on-rails-dd66ac17a1d0

**Pull request merged**